From 630b78d79e4e63d221b57726a3bc3855cedf4e28 Mon Sep 17 00:00:00 2001
From: Marius Wachtler <undingen@gmail.com>
Date: Tue, 22 Sep 2015 21:21:22 +0100
Subject: [PATCH] WIP: changes to make lxml work with pyston added a backport
 of a 3.1.2 patch in order to support the libxml2 travis-ci version + ubuntu
 14.04
 https://github.com/lxml/lxml/commit/3b661503609314af8ae1ba23501cca901e21cae3

---
 src/lxml/includes/etree_defs.h | 8 +++++++-
 src/lxml/includes/tree.pxd     | 3 +++
 src/lxml/serializer.pxi        | 5 +++--
 3 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/src/lxml/includes/etree_defs.h b/src/lxml/includes/etree_defs.h
index ff596ce..d3baef9 100644
--- a/src/lxml/includes/etree_defs.h
+++ b/src/lxml/includes/etree_defs.h
@@ -20,11 +20,15 @@
 #define va_int(ap)     va_arg(ap, int)
 #define va_charptr(ap) va_arg(ap, char *)
 
+/*
+Pyston change:
 #ifdef PYPY_VERSION
 #    define IS_PYPY 1
 #else
 #    define IS_PYPY 0
 #endif
+*/
+#define IS_PYPY 0
 
 #if PY_VERSION_HEX >= 0x03000000
 #  define IS_PYTHON3 1
@@ -152,7 +156,9 @@
 #ifndef LIBXML2_NEW_BUFFER
    typedef xmlBuffer xmlBuf;
 #  define xmlBufContent(buf) xmlBufferContent(buf)
-#  define xmlBufLength(buf) xmlBufferLength(buf)
+// Pyston change: change xmlBufLength to xmlBufUse. backport of lxml 3.1.2 patch
+//#  define xmlBufLength(buf) xmlBufferLength(buf)
+#  define xmlBufUse(buf) xmlBufferLength(buf)
 #endif
 
 /* libexslt 1.1.25+ support EXSLT functions in XPath */
diff --git a/src/lxml/includes/tree.pxd b/src/lxml/includes/tree.pxd
index afd49ba..a086384 100644
--- a/src/lxml/includes/tree.pxd
+++ b/src/lxml/includes/tree.pxd
@@ -363,6 +363,9 @@ cdef extern from "libxml/tree.h":
     cdef int xmlBufferLength(xmlBuffer* buf) nogil
     cdef const_xmlChar* xmlBufContent(xmlBuf* buf) nogil # new in libxml2 2.9
     cdef size_t xmlBufLength(xmlBuf* buf) nogil # new in libxml2 2.9
+    # Pyston change: change xmlBufLength to xmlBufUse. backport of lxml 3.1.2 patch
+    # cdef size_t xmlBufLength(xmlBuf* buf) nogil
+    cdef size_t xmlBufUse(xmlBuf* buf) nogil
     cdef int xmlKeepBlanksDefault(int val) nogil
     cdef xmlChar* xmlNodeGetBase(xmlDoc* doc, xmlNode* node) nogil
     cdef void xmlNodeSetBase(xmlNode* node, const_xmlChar* uri) nogil
diff --git a/src/lxml/serializer.pxi b/src/lxml/serializer.pxi
index f62887d..8579ed5 100644
--- a/src/lxml/serializer.pxi
+++ b/src/lxml/serializer.pxi
@@ -132,12 +132,13 @@ cdef _tostring(_Element element, encoding, doctype, method,
         _raiseSerialisationError(error_result)
 
     try:
+        # Pyston change: change xmlBufLength to xmlBufUse. backport of lxml 3.1.2 patch
         if encoding is _unicode:
             result = (<unsigned char*>tree.xmlBufContent(
-                c_result_buffer))[:tree.xmlBufLength(c_result_buffer)].decode('UTF-8')
+                c_result_buffer))[:tree.xmlBufUse(c_result_buffer)].decode('UTF-8')
         else:
             result = <bytes>(<unsigned char*>tree.xmlBufContent(
-                c_result_buffer))[:tree.xmlBufLength(c_result_buffer)]
+                c_result_buffer))[:tree.xmlBufUse(c_result_buffer)]
     finally:
         error_result = tree.xmlOutputBufferClose(c_buffer)
     if error_result < 0:
-- 
1.9.1

